### ISSUE TITLE
[Feature request] Library for matching nested expressions

### ISSUE THREAD

**Issue by RunDevelopment** [PrismJS/prism#2374]
**Motivation**
Javascript RegExp is unable to match nested structures (e.g. nested brackets) because they cannot be accepted by FSA. This is one of the main limitations Prism has as a regex-based syntax highlighter. 

Right now, we usually approximate such nested structures by creating long regexes that match support some maximum level of nesting and then fail. You can see this approach in languages such as JSX and C#.
However, this is neither pretty as the nested patterns as fairly complex, and to read and change, but it is also prone to errors because we then usually compromise between correctness and pattern length. When this compromise is unacceptable and a high level of nesting is necessary, the resulting pattern can become thousands of characters long (e.g. [this generated pattern](https://github.com/PrismJS/prism/blob/8a72830ab4b08d8d21ea4fc75fa27ba1ea872da2/components/prism-lilypond.js#L10)). The blowup in pattern length can even be exponential, so this solution really does not scale.

**Description**
Prism's matching algorithm doesn't actually require RegExp objects. The pattern only has to _behave_ like on. `matchGrammar` only ever uses the `exec` and `lastIndex` interface of RegExp objects, so if an object implements those two, Prism can use it. (ignoring the RegExp specific global check)

This means, we can make our own [PDA](https://en.wikipedia.org/wiki/Pushdown_automaton) implementation and it will just work!
Prism will then use the `exec` method and the PDA determines if and where the given string matches a pattern. This will be no different from what a normal regex does with the difference being that a PDA is more powerful automaton model. The PDA does not define the inside of a token; it is a drop-in replacement for normal regular expressions.

The implementation will __not be part of Core__ but will be a separate library like Markup templating instead. 
Core should stay simple and languages that don't require a PDA shouldn't have to pay for it. Luckily, this is very easy because of Prism's modular nature.

**Requirements**
The library has to be powerful enough to handle the following cases:

1. [Lilypond's Scheme expressions](https://github.com/PrismJS/prism/blob/8a72830ab4b08d8d21ea4fc75fa27ba1ea872da2/components/prism-lilypond.js#L3) (that can itself contain lilypond)
2. [JSX tags](https://github.com/PrismJS/prism/blob/8a72830ab4b08d8d21ea4fc75fa27ba1ea872da2/components/prism-jsx.js#L6)
3. [C# generics](https://github.com/PrismJS/prism/blob/8a72830ab4b08d8d21ea4fc75fa27ba1ea872da2/components/prism-csharp.js#L65)
4. [QML's JS expressions](https://github.com/PrismJS/prism/blob/8a72830ab4b08d8d21ea4fc75fa27ba1ea872da2/components/prism-qml.js#L6)
5. [JS template strings](https://github.com/PrismJS/prism/blob/8a72830ab4b08d8d21ea4fc75fa27ba1ea872da2/components/prism-javascript.js#L64)

Since we can't combine these PDAs with regexes, they have to be powerful enough to stand on their own. They have to be able to do everything normal regexes can do (maybe except for backreferences).

**Comment by joshgoebel** [comment]
> Prism's matching algorithm doesn't actually require RegExp objects. The pattern only has to behave like on. matchGrammar only ever uses the exec and lastIndex interface of RegExp objects

That's a rather ingenious hack. :-)

> Since we can't combine these PDAs with regexes, they have to be powerful enough to stand on their own. They have to be able to do everything normal regexes can do (maybe except for backreferences).

Why the wall of separation though?  Couldn't you even define your PDA using regex for the matchers?  Or it's possible I don't know anything about PDA and I'm just thinking of the idea _abstractly_ vs a very particular implementation (which PDA perhaps is).

IE, nested parens:

```
Prism.languages.ini= {
	'comment': /^[ \t]*[;#].*$/m,
        'expression': generatePDA({begin: /(/, end: /)/, allowRecursion: true})
```

Or even fallback to raw Prism highlighting for the "inner" content...

```
Prism.languages.ini= {
	'comment': /^[ \t]*[;#].*$/m,
        'cssString': generatePDA({begin:/CSS"/, end: /"/, inside: Prism.languages.css})
```

**Comment by RunDevelopment** [comment]
> Couldn't you even define your PDA using regex for the matches?

That's the idea but I don't really know how to. My basic idea was to give a BNF-like grammar and convert that into a deterministic PDA like this:

```js
PDA({ start: concat(/\(/, ref('start'), /\)/) })
```

Problems arise when you look at less trivial examples like JS template strings:

```js
PDA({
    'start': ref('template'),
    'template': concat(/`/, star(choice(/\\./, ref('interpolation'))), /`/),
    'interpolation': concat(/\$\{/, ref('js-expr'), /\}/),
    'js-expr': star(choice(
        /<normal JS string>/,
        /<comment>/,
        concat(/\{/, ref('js-expr'), /\}/,
        ref('template')
    )),
})
```

**Comment by joshgoebel** [comment]
> I don't really know how to this.

Is there a specific piece you're struggling with?

It looks like you're trying to build a tiny Highlight.js grammar inside there to me. :-)  Our engine is already very much like this (rules, nesting, recursion, etc).  So while you could this (and it might indeed be fun/useful) I feel like for anything non-trivial you're going to end up building quite a complex separate little parsing engine.  And things like choice will burn some cycles for sure.

It's possible a recursive solution might be simper than you'd think though.  We have a lot of legacy complexity regarding WHAT we do with matches (skip them, throw them back, conditional matches, one deep rule can cancel multiple levels fo the chain, etc)...  Might be much simpler if you just did the basic:

- find begin
- match a choice then recurse into that choice
- find end

But you add complexity with things like star (vs say single), etc. :-)  We don't have that (though there are different hacks).    All our rules are always `star` by default. :-)  But I fell `star` or numeric match counts might be easier than a lot of the things we do. :)  Since you're just limiting the internal loop.

**Comment by RunDevelopment** [comment]
> Is there a specific piece you're struggling with?

Nah, it's the general approach. This is a first for me, so I just have to read up on the literature and that's it. My problem is that I couldn't find a solution that is:

- Easy to implement (< 1k loc, doesn't require extensive analysis of the grammar)
- Easy to use (the grammar should be in a "natural" form)
- Fast. I want it to be fully determinist, so no backtracking. (Regexes within the solution are allowed to backtrack though)

This is probably one of those classic pick-2 situations but I'll continue to search for a good solution when I find the time. Rn there are more pressing issues IMO.

> It's possible a recursive solution might be simpler than you'd think though.

Recursion isn't possible because people might be able to construct a string that causes a stack overflow.

**Comment by joshgoebel** [comment]
> Easy to implement... < 1k loc... Easy to use... Fast... No recursion.

ROFL. I read that initially as `1kb of code`, not `1k LOC`. :-)  I thought you wanted to win a knife fight blindfolded with your hands tied behind your back and drunk... ;-)  1000 lines makes it a lot more doable.

Highlight.js "core" + "language compiler" is < 1k LOC, so I'd say it's doable. :-)   You just have to figure out how you want to structure your grammar.  We're not recursive though we may want to be in the future.  It would make backtracking easier to add (which sometimes I think we want and sometimes scares me to death).  Although I'm not sure how "fast" you mean when you say "fast"... I've always thought Highlight.js is pretty fast (for what it does), but it's possible when compared to something like Prism that we're a tortoise.  I don't know.  I'd love to see a comparison.

> Recursion isn't possible because people might be able to construct a string that causes a stack overflow.

It's been a while since I tested but the depth is like 1000-2000 or was it 10,000 to 20,000? I can't recall.  It's quite some code sample that would need to recurse that deeply.  If Highlight.js goes recursive in the future we'll just add a depth limit and say "hey we don't work perfectly for that 0.01% of edge cases". After all it's impossible to correct highlight 100% of cases anyways.  

So I feel like that may be a bit of an artificial restriction you're imposing on yourself. :-)

**Comment by joshgoebel** [comment]
I have some days where I think it'd be fun to implement the smallest (not fastest) HLJS "compatible" parser possible... but then there is no practical point, so I go back something more practical.  We're "small enough" for most uses and if someone truly needs smaller or tiny there is Prism.
